# Machine Learning With Spark ML
In this lab assignment, you will complete a project by going through the following steps:
1. Get the data.
2. Discover the data to gain insights.
3. Prepare the data for Machine Learning algorithms.
4. Select a model and train it.
5. Fine-tune your model.
6. Present your solution.

As a dataset, we use the California Housing Prices dataset from the StatLib repository. This dataset was based on data from the 1990 California census. The dataset has the following columns
1. `longitude`: a measure of how far west a house is (a higher value is farther west)
2. `latitude`: a measure of how far north a house is (a higher value is farther north)
3. `housing_,median_age`: median age of a house within a block (a lower number is a newer building)
4. `total_rooms`: total number of rooms within a block
5. `total_bedrooms`: total number of bedrooms within a block
6. `population`: total number of people residing within a block
7. `households`: total number of households, a group of people residing within a home unit, for a block
8. `median_income`: median income for households within a block of houses
9. `median_house_value`: median house value for households within a block
10. `ocean_proximity`: location of the house w.r.t ocean/sea

---
# 1. Get the data
Let's start the lab by loading the dataset. The can find the dataset at `data/housing.csv`. To infer column types automatically, when you are reading the file, you need to set `inferSchema` to true. Moreover enable the `header` option to read the columns' name from the file.

In [1]:
// TODO: Replace <FILL IN> with appropriate code

val housing = spark.read.format("csv").option("header", "true").option("inferSchema","true")load("data/housing.csv")

housing = [longitude: double, latitude: double ... 8 more fields]


[longitude: double, latitude: double ... 8 more fields]

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

---
# 2. Discover the data to gain insights
Now it is time to take a look at the data. In this step we are going to take a look at the data a few different ways:
* See the schema and dimension of the dataset
* Look at the data itself
* Statistical summary of the attributes
* Breakdown of the data by the categorical attribute variable
* Find the correlation among different attributes
* Make new attributes by combining existing attributes

## 2.1. Schema and dimension
Print the schema of the dataset

In [2]:
// TODO: Replace <FILL IN> with appropriate code

housing.show()

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|          452600.0|       NEAR BAY|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|          358500.0|       NEAR BAY|
|  -122.24|   37.85|              52.0|     1467.0|         190.0|     496.0|     177.0|       7.2574|          352100.0|       NEAR BAY|
|  -122.25|   37.85|              52.0|     1274.0|         235.0|     558.0|     219.0|       5.6431|          341300.0|       NEAR BAY|
|  -122.25|   37.85|              

Print the number of records in the dataset.

In [3]:
// TODO: Replace <FILL IN> with appropriate code

housing.count()

20640

## 2.2. Look at the data
Print the first five records of the dataset.

In [4]:
// TODO: Replace <FILL IN> with appropriate code

housing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|          452600.0|       NEAR BAY|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|          358500.0|       NEAR BAY|
|  -122.24|   37.85|              52.0|     1467.0|         190.0|     496.0|     177.0|       7.2574|          352100.0|       NEAR BAY|
|  -122.25|   37.85|              52.0|     1274.0|         235.0|     558.0|     219.0|       5.6431|          341300.0|       NEAR BAY|
|  -122.25|   37.85|              

Print the number of records with population more than 10000.

In [5]:
// TODO: Replace <FILL IN> with appropriate code

housing.where("population > 10000").count()

23

## 2.3. Statistical summary
Print a summary of the table statistics for the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. You can use the `describe` command.

In [6]:
// TODO: Replace <FILL IN> with appropriate code

val a=housing.select("housing_median_age","total_rooms","median_house_value","population").describe()
a.show()

+-------+------------------+------------------+------------------+------------------+
|summary|housing_median_age|       total_rooms|median_house_value|        population|
+-------+------------------+------------------+------------------+------------------+
|  count|             20640|             20640|             20640|             20640|
|   mean|28.639486434108527|2635.7630813953488|206855.81690891474|1425.4767441860465|
| stddev| 12.58555761211163|2181.6152515827944|115395.61587441359|  1132.46212176534|
|    min|               1.0|               2.0|           14999.0|               3.0|
|    max|              52.0|           39320.0|          500001.0|           35682.0|
+-------+------------------+------------------+------------------+------------------+



a = [summary: string, housing_median_age: string ... 3 more fields]


[summary: string, housing_median_age: string ... 3 more fields]

Print the maximum age (`housing_median_age`), the minimum number of rooms (`total_rooms`), and the average of house values (`median_house_value`).

In [7]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.sql.functions._

housing.select(max("housing_median_age"),min("total_rooms"),avg("median_house_value")).show()

+-----------------------+----------------+-----------------------+
|max(housing_median_age)|min(total_rooms)|avg(median_house_value)|
+-----------------------+----------------+-----------------------+
|                   52.0|             2.0|     206855.81690891474|
+-----------------------+----------------+-----------------------+



## 2.4. Breakdown the data by categorical data
Print the number of houses in different areas (`ocean_proximity`), and sort them in descending order.

In [8]:
// TODO: Replace <FILL IN> with appropriate code

housing.groupBy("ocean_proximity").agg(count("ocean_proximity")).sort(desc("ocean_proximity")).show()

+---------------+----------------------+
|ocean_proximity|count(ocean_proximity)|
+---------------+----------------------+
|     NEAR OCEAN|                  2658|
|       NEAR BAY|                  2290|
|         ISLAND|                     5|
|         INLAND|                  6551|
|      <1H OCEAN|                  9136|
+---------------+----------------------+



Print the average value of the houses (`median_house_value`) in different areas (`ocean_proximity`), and call the new column `avg_value` when print it.

In [9]:
// TODO: Replace <FILL IN> with appropriate code

housing.groupBy("ocean_proximity").agg(avg("median_house_value") as "avg_vale").sort(desc("ocean_proximity")).show()

+---------------+------------------+
|ocean_proximity|          avg_vale|
+---------------+------------------+
|     NEAR OCEAN|249433.97742663656|
|       NEAR BAY|259212.31179039303|
|         ISLAND|          380440.0|
|         INLAND|124805.39200122119|
|      <1H OCEAN|240084.28546409807|
+---------------+------------------+



Rewrite the above question in SQL.

In [10]:
// TODO: Replace <FILL IN> with appropriate code

housing.createOrReplaceTempView("df")
spark.sql("select ocean_proximity,avg(median_house_value) as avg_value from df group by ocean_proximity order by ocean_proximity desc ").show()

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|     NEAR OCEAN|249433.97742663656|
|       NEAR BAY|259212.31179039303|
|         ISLAND|          380440.0|
|         INLAND|124805.39200122119|
|      <1H OCEAN|240084.28546409807|
+---------------+------------------+



## 2.5. Correlation among attributes
Print the correlation among the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. To do so, first you need to put these attributes into one vector. Then, compute the standard correlation coefficient (Pearson) between every pair of attributes in this new vector. To make a vector of these attributes, you can use the `VectorAssembler` Transformer.

In [11]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.VectorAssembler

val va = new VectorAssembler().setInputCols(Array("housing_median_age","total_rooms","median_house_value")).setOutputCol("features")

val housingAttrs = va.transform(housing)

housingAttrs.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+--------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|            features|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+--------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|          452600.0|       NEAR BAY|[41.0,880.0,45260...|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|          358500.0|       NEAR BAY|[21.0,7099.0,3585...|
|  -122.24|   37.85|              52.0|     1467.0|         190.0|     496.0|     177.0|       7.2574|          352100.0|       NEAR BAY|[52.0,1467.0,3521...|
|  -122.25|   37.85|              52.0|     12

va = vecAssembler_3d05504a1288
housingAttrs = [longitude: double, latitude: double ... 9 more fields]


[longitude: double, latitude: double ... 9 more fields]

In [12]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.linalg.Matrix
import org.apache.spark.ml.stat.Correlation
import org.apache.spark.sql.Row

val Row(coeff: Matrix) = Correlation.corr(housingAttrs,"features").head()

println(s"The standard correlation coefficient:\n ${coeff}")

The standard correlation coefficient:
 1.0                   -0.36126220122231784  0.10562341249318154  
-0.36126220122231784  1.0                   0.13415311380654338  
0.10562341249318154   0.13415311380654338   1.0                  


coeff = 


1.0                   -0.36126220122231784  0.10562341249318154
-0.36126220122231784  1.0                   0.13415311380654338
0.10562341249318154   0.13415311380654338   1.0


## 2.6. Combine and make new attributes
Now, let's try out various attribute combinations. In the given dataset, the total number of rooms in a block is not very useful, if we don't know how many households there are. What we really want is the number of rooms per household. Similarly, the total number of bedrooms by itself is not very useful, and we want to compare it to the number of rooms. And the population per household seems like also an interesting attribute combination to look at. To do so, add the three new columns to the dataset as below. We will call the new dataset the `housingExtra`.
```
rooms_per_household = total_rooms / households
bedrooms_per_room = total_bedrooms / total_rooms
population_per_household = population / households
```

rooms_per_household = total_rooms / households;
bedrooms_per_room = total_bedrooms / total_rooms;
population_per_household = population / households;

In [13]:
// TODO: Replace <FILL IN> with appropriate code

val housingCol1 = housing.withColumn("rooms_per_household",expr("total_rooms/households"))
val housingCol2 = housingCol1.withColumn("bedrooms_per_room",expr("total_bedrooms/total_rooms"))
val housingExtra = housingCol2.withColumn("population_per_household",expr("population/households"))

housingExtra.select("rooms_per_household", "bedrooms_per_room", "population_per_household").show(5)

+-------------------+-------------------+------------------------+
|rooms_per_household|  bedrooms_per_room|population_per_household|
+-------------------+-------------------+------------------------+
|  6.984126984126984|0.14659090909090908|      2.5555555555555554|
|  6.238137082601054|0.15579659106916466|       2.109841827768014|
|  8.288135593220339|0.12951601908657123|      2.8022598870056497|
| 5.8173515981735155|0.18445839874411302|       2.547945205479452|
|  6.281853281853282| 0.1720958819913952|      2.1814671814671813|
+-------------------+-------------------+------------------------+
only showing top 5 rows



housingCol1 = [longitude: double, latitude: double ... 9 more fields]
housingCol2 = [longitude: double, latitude: double ... 10 more fields]
housingExtra = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

---
## 3. Prepare the data for Machine Learning algorithms
Before going through the Machine Learning steps, let's first rename the label column from `median_house_value` to `label`.

In [14]:
// TODO: Replace <FILL IN> with appropriate code

val renamedHousing = housingExtra.withColumnRenamed("median_house_value","label");

renamedHousing = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

Now, we want to separate the numerical attributes from the categorical attribute (`ocean_proximity`) and keep their column names in two different lists. Moreover, sice we don't want to apply the same transformations to the predictors (features) and the label, we should remove the label attribute from the list of predictors. 

In [15]:
// label columns
val colLabel = "label"

// categorical columns
val colCat = "ocean_proximity"

// numerical columns
val colNum = renamedHousing.columns.filter(_ != colLabel).filter(_ != colCat)

colLabel = label
colCat = ocean_proximity
colNum = Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household)


Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household)

## 3.1. Prepare continuse attributes
### Data cleaning
Most Machine Learning algorithms cannot work with missing features, so we should take care of them. As a first step, let's find the columns with missing values in the numerical attributes. To do so, we can print the number of missing values of each continues attributes, listed in `colNum`.

In [16]:
// TODO: Replace <FILL IN> with appropriate code

for (c <- colNum) {
    println(c)
   val a =renamedHousing.where(renamedHousing.col(c).isNull).count()

   println(a) 
}

longitude
0
latitude
0
housing_median_age
0
total_rooms
0
total_bedrooms
207
population
0
households
0
median_income
0
rooms_per_household
0
bedrooms_per_room
207
population_per_household
0


As we observerd above, the `total_bedrooms` and `bedrooms_per_room` attributes have some missing values. One way to take care of missing values is to use the `Imputer` Transformer, which completes missing values in a dataset, either using the mean or the median of the columns in which the missing values are located. To use it, you need to create an `Imputer` instance, specifying that you want to replace each attribute's missing values with the "median" of that attribute.

In [17]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.Imputer

val imputer = new Imputer().setStrategy("median").setInputCols(Array("total_bedrooms","bedrooms_per_room")).setOutputCols(Array("total_bedrooms","bedrooms_per_room"))                                    
val imputedHousing = imputer.fit(renamedHousing).transform(renamedHousing)

imputedHousing.select("total_bedrooms", "bedrooms_per_room").show(5)

+--------------+-------------------+
|total_bedrooms|  bedrooms_per_room|
+--------------+-------------------+
|         129.0|0.14659090909090908|
|        1106.0|0.15579659106916466|
|         190.0|0.12951601908657123|
|         235.0|0.18445839874411302|
|         280.0| 0.1720958819913952|
+--------------+-------------------+
only showing top 5 rows



imputer = imputer_8383f37ffc97
imputedHousing = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

### Scaling
One of the most important transformations you need to apply to your data is feature scaling. With few exceptions, Machine Learning algorithms don't perform well when the input numerical attributes have very different scales. This is the case for the housing data: the total number of rooms ranges from about 6 to 39,320, while the median incomes only range from 0 to 15. Note that scaling the label attribues is generally not required.

One way to get all attributes to have the same scale is to use standardization. In standardization, for each value, first it subtracts the mean value (so standardized values always have a zero mean), and then it divides by the variance so that the resulting distribution has unit variance. To do this, we can use the `StandardScaler` Estimator. To use `StandardScaler`, again we need to convert all the numerical attributes into a big vectore of features using `VectorAssembler`, and then call `StandardScaler` on that vactor.

In [18]:
imputedHousing.columns

Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, label, ocean_proximity, rooms_per_household, bedrooms_per_room, population_per_household)

In [19]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}

val va = new VectorAssembler().setInputCols(colNum).setOutputCol("featuresFlat")
val featuredHousing = va.transform(imputedHousing)

val scaler = new StandardScaler().setInputCol("featuresFlat").setOutputCol("scalar")
val scaledHousing = scaler.fit(featuredHousing).transform(featuredHousing)

scaledHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|        featuresFlat|              scalar|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|[-122.23,37.88,41...|[-61.007269596069...|
|  -122.22|   37.86|              21.0|     7099.0|        1

va = vecAssembler_fc5867a69adf
featuredHousing = [longitude: double, latitude: double ... 12 more fields]
scaler = stdScal_082c13772a7c
scaledHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

## 3.2. Prepare categorical attributes
After imputing and scaling the continuse attributes, we should take care of the categorical attributes. Let's first print the number of distict values of the categirical attribute `ocean_proximity`.

In [20]:
// TODO: Replace <FILL IN> with appropriate code

renamedHousing.select("ocean_proximity").distinct().count()

5

### String indexer
Most Machine Learning algorithms prefer to work with numbers. So let's convert the categorical attribute `ocean_proximity` to numbers. To do so, we can use the `StringIndexer` that encodes a string column of labels to a column of label indices. The indices are in [0, numLabels), ordered by label frequencies, so the most frequent label gets index 0.

In [21]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.StringIndexer

val indexer = new StringIndexer().setInputCol(colCat).setOutputCol("catInd")
val idxHousing = indexer.fit(renamedHousing).transform(renamedHousing)

idxHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|catInd|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|   3.0|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|358500.0|       NEAR BAY|  6.238137082601054|0.15579659106916466|       2.109841827768014|   3.0|


indexer = strIdx_6890ef73a1e9
idxHousing = [longitude: double, latitude: double ... 12 more fields]


[longitude: double, latitude: double ... 12 more fields]

Now we can use this numerical data in any Machine Learning algorithm. You can look at the mapping that this encoder has learned using the `labels` method: "<1H OCEAN" is mapped to 0, "INLAND" is mapped to 1, etc.

In [22]:
indexer.fit(renamedHousing).labels

Array(<1H OCEAN, INLAND, NEAR OCEAN, NEAR BAY, ISLAND)

### One-hot encoding
Now, convert the label indices built in the last step into one-hot vectors. To do this, you can take advantage of the `OneHotEncoderEstimator` Estimator.

In [23]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.OneHotEncoderEstimator



val encoder = new OneHotEncoderEstimator().setInputCols(Array("catInd")).setOutputCols(Array("catEncode"))
val ohHousing = encoder.fit(idxHousing).transform(idxHousing)

ohHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+------+-------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|catInd|    catEncode|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+------+-------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|   3.0|(4,[3],[1.0])|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|358500.0|       NEAR BAY|  6.2381370826010

encoder = oneHotEncoder_bb7f98f15112
ohHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

---
# 4. Pipeline
As you can see, there are many data transformation steps that need to be executed in the right order. For example, you called the `Imputer`, `VectorAssembler`, and `StandardScaler` from left to right. However, we can use the `Pipeline` class to define a sequence of Transformers/Estimators, and run them in order. A `Pipeline` is an `Estimator`, thus, after a Pipeline's `fit()` method runs, it produces a `PipelineModel`, which is a `Transformer`.

Now, let's create a pipeline called `numPipeline` to call the numerical transformers you built above (`imputer`, `va`, and `scaler`) in the right order from left to right, as well as a pipeline called `catPipeline` to call the categorical transformers (`indexer` and `encoder`). Then, put these two pipelines `numPipeline` and `catPipeline` into one pipeline.

In [24]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.{Pipeline, PipelineModel, PipelineStage}

val numPipeline = new Pipeline().setStages(Array(imputer, va,scaler))
val catPipeline = new Pipeline().setStages(Array(indexer, encoder))
val pipeline = new Pipeline().setStages(Array(numPipeline, catPipeline))
val newHousing = pipeline.fit(renamedHousing).transform(renamedHousing)

newHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+------+-------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|        featuresFlat|              scalar|catInd|    catEncode|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+------+-------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|[-122.23,37.88,41...|[-61.007269596069..

numPipeline = pipeline_9f6d5e7089c5
catPipeline = pipeline_7e277e2ac6c6
pipeline = pipeline_3e2fc1665622
newHousing = [longitude: double, latitude: double ... 15 more fields]


[longitude: double, latitude: double ... 15 more fields]

Now, use `VectorAssembler` to put all attributes of the final dataset `newHousing` into a big vector, and call the new column `features`.

In [25]:
// TODO: Replace <FILL IN> with appropriate code

//val cols = newHousing.columns.filter(_ != colCat)
val cols=Array("scalar","catEncode")
println(cols);
val va2 = new VectorAssembler().setInputCols(cols).setOutputCol("features")
val dataset1 = va2.transform(newHousing).select("features", "label")

dataset1.show(5)

[Ljava.lang.String;@7f977d8d
+--------------------+--------+
|            features|   label|
+--------------------+--------+
|[-61.007269596069...|452600.0|
|[-61.002278409814...|358500.0|
|[-61.012260782324...|352100.0|
|[-61.017251968579...|341300.0|
|[-61.017251968579...|342200.0|
+--------------------+--------+
only showing top 5 rows



cols = Array(scalar, catEncode)
va2 = vecAssembler_ee688e20ead3
dataset1 = [features: vector, label: double]


[features: vector, label: double]

---
# 5. Make a model
Here we going to make four different regression models:
* Linear regression model
* Decission tree regression
* Random forest regression
* Gradient-booster forest regression

But, before giving the data to train a Machine Learning model, let's first split the data into training dataset (`trainSet`) with 80% of the whole data, and test dataset (`testSet`) with 20% of it.

In [26]:
// TODO: Replace <FILL IN> with appropriate code
val Array(trainSet1, testSet1) = dataset1.randomSplit(Array(0.8, 0.2))

testSet1.show()

+--------------------+--------+
|            features|   label|
+--------------------+--------+
|[-62.005506847089...| 73200.0|
|[-61.995524474579...| 66900.0|
|[-61.985542102068...| 74600.0|
|[-61.980550915813...|107000.0|
|[-61.980550915813...| 72200.0|
|[-61.980550915813...| 70200.0|
|[-61.980550915813...| 70500.0|
|[-61.975559729558...| 60000.0|
|[-61.975559729558...| 75500.0|
|[-61.970568543303...| 85100.0|
|[-61.970568543303...| 85600.0|
|[-61.970568543303...| 96000.0|
|[-61.965577357048...|100600.0|
|[-61.960586170793...| 75100.0|
|[-61.960586170793...| 99600.0|
|[-61.960586170793...|143400.0|
|[-61.960586170793...|122400.0|
|[-61.955594984538...| 92800.0|
|[-61.945612612028...| 96100.0|
|[-61.945612612028...| 90200.0|
+--------------------+--------+
only showing top 20 rows



trainSet1 = [features: vector, label: double]
testSet1 = [features: vector, label: double]


[features: vector, label: double]

## 5.1. Linear regression model
Now, train a Linear Regression model using the `LinearRegression` class. Then, print the coefficients and intercept of the model, as well as the summary of the model over the training set by calling the `summary` method.

In [27]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.LinearRegression

// train the model
val lr1 = new LinearRegression().setMaxIter(10)
.setFeaturesCol("features").setLabelCol("label").setSolver("normal")
val lrModel1 = lr1.fit(trainSet1)
val trainingSummary = lrModel1.summary

println(s"Coefficients: ${lrModel1.coefficients}, Intercept: ${lrModel1.intercept}")
println(s"RMSE: ${trainingSummary.rootMeanSquaredError}")

Coefficients: [-55546.56043266049,-56452.04214475486,13692.654475868036,7241.058647006913,2937.7422557907666,-46634.74195692103,41227.15737469756,77970.63361154216,6103.1243180291995,15952.731501948629,750.5776374053479,-166119.78762122296,-200071.68439523797,-161342.38161209383,-170738.32659362792], Intercept: -2260083.9871313805
RMSE: 67476.07036192763


lr1 = linReg_44cc73930bcd
lrModel1 = linReg_44cc73930bcd
trainingSummary = org.apache.spark.ml.regression.LinearRegressionTrainingSummary@9b6efac


org.apache.spark.ml.regression.LinearRegressionTrainingSummary@9b6efac

Now, use `RegressionEvaluator` to measure the root-mean-square-erroe (RMSE) of the model on the test dataset.

In [28]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.evaluation.RegressionEvaluator

// make predictions on the test data
val predictions = lrModel1.transform(testSet1)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator().setMetricName("rmse")
.setPredictionCol("prediction")
.setLabelCol("label")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|114636.98888270231| 73200.0|[-62.005506847089...|
|107105.56739898026| 66900.0|[-61.995524474579...|
| 93955.01353170304| 74600.0|[-61.985542102068...|
|180167.20266358973|107000.0|[-61.980550915813...|
|154537.76002162974| 72200.0|[-61.980550915813...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 69537.35484194856


predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_9b374f0f7486
rmse = 69537.35484194856


69537.35484194856

## 5.2. Decision tree regression
Repeat what you have done on Regression Model to build a Decision Tree model. Use the `DecisionTreeRegressor` to make a model and then measure its RMSE on the test dataset.

In [29]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.DecisionTreeRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val dt = new DecisionTreeRegressor().setLabelCol("label").setFeaturesCol("features")

// train the model
val dtModel = dt.fit(trainSet1)

// make predictions on the test data
val predictions = dtModel.transform(testSet1)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setMetricName("rmse")

val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|165532.87893175075| 73200.0|[-62.005506847089...|
|137563.56144067796| 66900.0|[-61.995524474579...|
|147028.97959183675| 74600.0|[-61.985542102068...|
|165532.87893175075|107000.0|[-61.980550915813...|
|165532.87893175075| 72200.0|[-61.980550915813...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 68421.60767593021


dt = dtr_985acf79c571
dtModel = DecisionTreeRegressionModel (uid=dtr_985acf79c571) of depth 5 with 63 nodes
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_0cdd02bb529d
rmse = 68421.60767593021


68421.60767593021

## 5.3. Random forest regression
Let's try the test error on a Random Forest Model. Youcan use the `RandomForestRegressor` to make a Random Forest model.

In [30]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.RandomForestRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val rf = new RandomForestRegressor().setLabelCol("label").setFeaturesCol("features")

// train the model
val rfModel = rf.fit(trainSet1)

// make predictions on the test data
val predictions = rfModel.transform(testSet1)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setMetricName("rmse")
.setPredictionCol("prediction")
.setLabelCol("label")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 172967.7321409177| 73200.0|[-62.005506847089...|
|145430.12284887998| 66900.0|[-61.995524474579...|
| 147053.0859961635| 74600.0|[-61.985542102068...|
|172519.32045489136|107000.0|[-61.980550915813...|
|147418.34493311186| 72200.0|[-61.980550915813...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 66104.413007993


rf = rfr_305a748d4d0b
rfModel = RandomForestRegressionModel (uid=rfr_305a748d4d0b) with 20 trees
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_44514d253545
rmse = 66104.413007993


66104.413007993

## 5.4. Gradient-boosted tree regression
Fianlly, we want to build a Gradient-boosted Tree Regression model and test the RMSE of the test data. Use the `GBTRegressor` to build the model.

In [31]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.GBTRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val gb = new GBTRegressor().setLabelCol("label").setFeaturesCol("features")

// train the model
val gbModel = gb.fit(trainSet1)

// make predictions on the test data
val predictions = gbModel.transform(testSet1)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setMetricName("rmse")
.setPredictionCol("prediction")
.setLabelCol("label")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|131338.82462455294| 73200.0|[-62.005506847089...|
|105628.86788609039| 66900.0|[-61.995524474579...|
| 71719.27815297342| 74600.0|[-61.985542102068...|
|103366.86617259108|107000.0|[-61.980550915813...|
| 98961.09540183515| 72200.0|[-61.980550915813...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 55542.62529156017


gb = gbtr_9cc222e6590c
gbModel = GBTRegressionModel (uid=gbtr_9cc222e6590c) with 20 trees
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_211fcf327b83
rmse = 55542.62529156017


55542.62529156017

---
# 6. Hyperparameter tuning
An important task in Machie Learning is model selection, or using data to find the best model or parameters for a given task. This is also called tuning. Tuning may be done for individual Estimators such as LinearRegression, or for entire Pipelines which include multiple algorithms, featurization, and other steps. Users can tune an entire Pipeline at once, rather than tuning each element in the Pipeline separately. MLlib supports model selection tools, such as `CrossValidator`. These tools require the following items:
* Estimator: algorithm or Pipeline to tune (`setEstimator`)
* Set of ParamMaps: parameters to choose from, sometimes called a "parameter grid" to search over (`setEstimatorParamMaps`)
* Evaluator: metric to measure how well a fitted Model does on held-out test data (`setEvaluator`)

`CrossValidator` begins by splitting the dataset into a set of folds, which are used as separate training and test datasets. For example with `k=3` folds, `CrossValidator` will generate 3 (training, test) dataset pairs, each of which uses 2/3 of the data for training and 1/3 for testing. To evaluate a particular `ParamMap`, `CrossValidator` computes the average evaluation metric for the 3 Models produced by fitting the Estimator on the 3 different (training, test) dataset pairs. After identifying the best `ParamMap`, `CrossValidator` finally re-fits the Estimator using the best ParamMap and the entire dataset.

Below, use the `CrossValidator` to select the best Random Forest model. To do so, you need to define a grid of parameters. Let's say we want to do the search among the different number of trees (1, 5, and 10), and different tree depth (5, 10, and 15).

In [32]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.RegressionEvaluator
import org.apache.spark.ml.tuning.CrossValidator

val paramGrid = new ParamGridBuilder().addGrid(rf.numTrees, Array(1,5,10))
    .addGrid(rf.maxDepth, Array(5,10,15))
    .build()

val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val cv = new CrossValidator().setEstimator(lr1)
    .setEvaluator(new RegressionEvaluator())
    .setEstimatorParamMaps(paramGrid)
    .setNumFolds(3)
val cvModel = cv.fit(trainSet1)

val predictions = cvModel.transform(testSet1)
predictions.select("prediction", "label", "features").show(5)

val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|114636.98888270231| 73200.0|[-62.005506847089...|
|107105.56739898026| 66900.0|[-61.995524474579...|
| 93955.01353170304| 74600.0|[-61.985542102068...|
|180167.20266358973|107000.0|[-61.980550915813...|
|154537.76002162974| 72200.0|[-61.980550915813...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 69537.35484194856


paramGrid = 


Array({
	rfr_305a748d4d0b-maxDepth: 5,
	rfr_305a748d4d0b-numTrees: 1
}, {
	rfr_305a748d4d0b-maxDepth: 5,
	rfr_305a748d4d0b-numTrees: 5
}, {
	rfr_305a748d4d0b-maxDepth: 5,
	rfr_305a748d4d0b-numTrees: 10
}, {
	rfr_305a748d4d0b-maxDepth: 10,
	rfr_305a748d4d0b-numTrees: 1
}, {
	rfr_305a748d4d0b-maxDepth: 10,
	rfr_305a748d4d0b-numTrees: 5
}, {
	rfr_305a748d4d0b-maxDepth: 10,
	rfr_305a748d4d0b-numTrees: 10
}, {
	rfr_305a748d4d0b-maxDepth: 15,
	rfr_305a748d4d0b-numTrees: 1
}, {
	rfr_305a748d4d0b-maxDepth: 15,
	rfr_305a748d4d0b-numTrees: 5
}, {
	rfr_305a748d4d0b-maxDepth: 15,
	rfr_305a...


---
# 7. Custom transformer
At the end of part two, we added extra columns to the `housing` dataset. Here, we are going to implement a Transformer to do the same task. The Transformer should take the name of two input columns `inputCol1` and `inputCol2`, as well as the name of ouput column `outputCol`. It, then, computes `inputCol1` divided by `inputCol2`, and adds its result as a new column to the dataset. The details of the implemeting a custom Tranfomer is explained [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types). Please read it before before starting to implement it.

First, define the given parameters of the Transformer and implement a method to validate their schemas (`StructType`).

In [33]:
import org.apache.spark.sql.types.{StructField, StructType, DoubleType}
import org.apache.spark.ml.param.{ParamMap, Param, Params}

trait MyParams extends Params {
    final val inputCol1 = new Param[String](this, "inputCol1","input colum1")
    final val inputCol2 = new Param[String](this, "inputCol2","input colum2")
    final val outputCol = new Param[String](this, "outputCol","output col")
    
  protected def validateAndTransformSchema(schema: StructType): StructType = {
       val idx = schema.fieldIndex($(inputCol1))
      val idx2 = schema.fieldIndex($(inputCol2))
    val field = schema.fields(idx)
      val field2 = schema.fields(idx2)
    if (field.dataType != DoubleType || field2.dataType != DoubleType) {
      throw new Exception(s"Input type ${field.dataType} did not match input type DoubleType")
    }
    // Add the return field
    schema.add(StructField($(outputCol), DoubleType, false))
      
   
      
  
  }
}


defined trait MyParams


Then, extend the class `Transformer`, and implement its setter functions for the input and output columns, and call then `setInputCol1`, `setInputCol2`, and `setOutputCol`. Morever, you need to override the methods `copy`, `transformSchema`, and the `transform`. The details of what you need to cover in these methods is given [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types).

In [34]:
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.ml.Transformer
import org.apache.spark.ml.param.{ParamMap, Param, Params}
import org.apache.spark.sql.{DataFrame, Dataset}
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.functions.{col, udf}

class MyTransformer(override val uid: String) extends Transformer with MyParams {
    def this() = this(Identifiable.randomUID("configurablewordcount"))
    
    def setInputCol1(value: String): this.type=set(inputCol1,value)
    
    def setInputCol2(value: String): this.type=set(inputCol2,value)
    
    def setOutputCol(value: String): this.type=set(outputCol,value)

    override def copy(extra: ParamMap) :MyTransformer=defaultCopy(extra)

    override def transformSchema(schema: StructType) :StructType={
        this.validateAndTransformSchema(schema: StructType)
    }
    
    override def transform(dataset: Dataset[_]): DataFrame = {
       val division = udf { (in1: Double,in2 :Double) => in1/in2 }
   
    dataset.withColumn($(outputCol), division(dataset.col($(inputCol1)),dataset.col($(inputCol2))) )    
        

  }
}

defined class MyTransformer


Now, an instance of `MyTransformer`, and set the input columns `total_rooms` and `households`, and the output column `rooms_per_household` and run it over the `housing` dataset.

In [35]:
val myTransformer = new MyTransformer().setInputCol1("total_rooms").setInputCol2("households").setOutputCol("rooms_per_household")

val myDataset = myTransformer.transform(housing).select("rooms_per_household").show(5)

+-------------------+
|rooms_per_household|
+-------------------+
|  6.984126984126984|
|  6.238137082601054|
|  8.288135593220339|
| 5.8173515981735155|
|  6.281853281853282|
+-------------------+
only showing top 5 rows



myTransformer = configurablewordcount_6eb1416e3139


myDataset: Unit = ()


configurablewordcount_6eb1416e3139

---
# 8. Custom estimator (predictor)
Now, it's time to implement your own linear regression with gradient descent algorithm as a brand new Estimator. The whole code of the Estimator is given to you, and you do not need to implement anything. It is just a sample that shows how to build a custom Estimator.

The gradient descent update for linear regression is:
$$
w_{i+1} = w_{i} - \alpha_{i} \sum\limits_{j=1}^n (w_i^\top x_j - y_j)x_j
$$

where $i$ is the iteration number of the gradient descent algorithm, and $j$ identifies the observation. Here, $w$ represents an array of weights that is the same size as the array of features and provides a weight for each of the features when finally computing the label prediction in the form:

$$
prediction = w^\top \cdot\ x
$$

where $w$ is the final array of weights computed by the gradient descent, $x$ is the array of features of the observation point and $prediction$ is the label we predict should be associated to this observation.

The given `Helper` class implements the helper methods:
* `dot`: implements the dot product of two vectors and the dot product of a vector and a scalar
* `sum`: implements addition of two vectors
* `fill`: creates a vector of predefined size and initialize it with the predefined value

What you need to do is to implement the methods of the Linear Regresstion class `LR`, which are
* `rmsd`: computes the Root Mean Square Error of a given RDD of tuples of (label, prediction) using the formula:
$$
rmse = \sqrt{\frac{\sum\limits_{i=1}^n (label - prediction)^2}{n}}
$$
* `gradientSummand`: computes the following formula:
$$
gs_{ij} = (w_i^\top x_j - y_j)x_j
$$
* `gradient`: computes the following formula:
$$
gradient = \sum\limits_{j=1}^n gs_{ij}
$$

In [36]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

case class Instance(label: Double, features: Vector)

object Helper extends Serializable {
  def dot(v1: Vector, v2: Vector): Double = {
    val m = Matrices.dense(1, v1.size, v1.toArray)
    m.multiply(v2).values(0)
  }

  def dot(v: Vector, s: Double): Vector = {
    val baseArray = v.toArray.map(vi => vi * s)
    Vectors.dense(baseArray)
  }

  def sumVectors(v1: Vector, v2: Vector): Vector = {
    val baseArray = ((v1.toArray) zip (v2.toArray)).map { case (val1, val2) => val1 + val2 }
    Vectors.dense(baseArray)
  }

  def fillVector(size: Int, fillVal: Double): Vector = Vectors.dense(Array.fill[Double](size)(fillVal));
}

defined class Instance
defined object Helper


In [37]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}




class LR() extends Serializable {
  def calcRMSE(labelsAndPreds: RDD[(Double, Double)]): Double = {
    val regressionMetrics = new RegressionMetrics(labelsAndPreds)
    regressionMetrics.rootMeanSquaredError
  }
  
  def gradientSummand(weights: Vector, lp: Instance): Vector = {
    val mult = (Helper.dot(weights, lp.features) - lp.label)
    val seq = (0 to lp.features.size - 1).map(i => lp.features(i) * mult)
    return Vectors.dense(seq.toArray)
  }
  
  def linregGradientDescent(trainData: RDD[Instance], numIters: Int): (Vector, Array[Double]) = {
    val n = trainData.count()
    val d = trainData.take(1)(0).features.size
    var w = Helper.fillVector(d, 0)
    val alpha = 1.0
    val errorTrain = Array.fill[Double](numIters)(0.0)

    for (i <- 0 until numIters) {
      val labelsAndPredsTrain = trainData.map(lp => (lp.label, Helper.dot(w, lp.features)))
      errorTrain(i) = calcRMSE(labelsAndPredsTrain)

      val gradient = trainData.map(lp => gradientSummand(w, lp)).reduce((v1, v2) => Helper.sumVectors(v1, v2))
      val alpha_i = alpha / (n * scala.math.sqrt(i + 1))
      val wAux = Helper.dot(gradient, (-1) * alpha_i)
      w = Helper.sumVectors(w, wAux)
    }
    (w, errorTrain)
  }
}

defined class LR


In [38]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}


abstract class MyLinearModel[FeaturesType, Model <: MyLinearModel[FeaturesType, Model]]
  extends PredictionModel[FeaturesType, Model] {
}

class MyLinearModelImpl(override val uid: String, val weights: Vector, val trainingError: Array[Double])
    extends MyLinearModel[Vector, MyLinearModelImpl] {

  override def copy(extra: ParamMap): MyLinearModelImpl = defaultCopy(extra)

  def predict(features: Vector): Double = {
    println("Predicting")
    val prediction = Helper.dot(weights, features)
    prediction
  }
}

defined class MyLinearModel
defined class MyLinearModelImpl


In [39]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

abstract class MyLinearRegression[
    FeaturesType,
    Learner <: MyLinearRegression[FeaturesType, Learner, Model],
    Model <: MyLinearModel[FeaturesType, Model]]
  extends Predictor[FeaturesType, Learner, Model] {
}

class MyLinearRegressionImpl(override val uid: String)
    extends MyLinearRegression[Vector, MyLinearRegressionImpl, MyLinearModelImpl] {
  def this() = this(Identifiable.randomUID("linReg"))

  override def copy(extra: ParamMap): MyLinearRegressionImpl = defaultCopy(extra)
  
  def train(dataset: Dataset[_]): MyLinearModelImpl = {
    println("Training")

    val numIters = 10

    val instances: RDD[Instance] = dataset.select(
      col($(labelCol)), col($(featuresCol))).rdd.map {
        case Row(label: Double, features: Vector) =>
          Instance(label, features)
      }

    val (weights, trainingError) = new LR().linregGradientDescent(instances, numIters)

    new MyLinearModelImpl(uid, weights, trainingError)
  }
}

defined class MyLinearRegression
defined class MyLinearRegressionImpl


In [74]:
import org.apache.spark.ml.evaluation.RegressionEvaluator

val lr = new MyLinearRegressionImpl().setLabelCol("label").setFeaturesCol("features")
val model = lr.fit(trainSet1)
val predictions = model.transform(testSet1)
predictions.select("prediction", "label", "features").show(5)

// Select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

Training
+--------------------+--------+--------------------+
|          prediction|   label|            features|
+--------------------+--------+--------------------+
|-8.61613519035840...| 73200.0|[-62.005506847089...|
|-8.62150999011452...| 66900.0|[-61.995524474579...|
|-8.61747570123055...| 74600.0|[-61.985542102068...|
|-8.59161184468638...|107000.0|[-61.980550915813...|
|-8.59471883851774E37| 72200.0|[-61.980550915813...|
+--------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 8.235481859085347E37


lr = linReg_6dd5a7674c56
model = linReg_6dd5a7674c56
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_8c0f7a82a52a
rmse = 8.235481859085347E37


8.235481859085347E37

---
# 9. An End-to-End Classification Test
As the last step, you are given a dataset called `data/ccdefault.csv`. The dataset represents default of credit card clients. It has 30,000 cases and 24 different attributes. More details about the dataset is available at `data/ccdefault.txt`. In this task you should make three models, compare their results and conclude the ideal solution. Here are the suggested steps:
1. Load the data.
2. Carry out some exploratory analyses (e.g., how various features and the target variable are distributed).
3. Train a model to predict the target variable (risk of `default`).
  - Employ three different models (logistic regression, decision tree, and random forest).
  - Compare the models' performances (e.g., AUC).
  - Defend your choice of best model (e.g., what are the strength and weaknesses of each of these models?).
4. What more would you do with this data? Anything to help you devise a better solution?

In [41]:
val data = spark.read.format("csv").option("header", "true").option("inferSchema","true")load("data/ccdefault.csv")

data = [ID: int, LIMIT_BAL: int ... 23 more fields]


[ID: int, LIMIT_BAL: int ... 23 more fields]

In [42]:
data.count()

30000

In [43]:
data.show(10)

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|DEFAULT|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0|      1|
|  2|   120000|  2|        2|       2| 26|   -1|    2|    0|    0|    0|    2|     2682|     1725|     2682|     3272|     3455|     3261|       0|    1000|    1000|    1000|       0|    2000|    

In [44]:
data.columns

Array(ID, LIMIT_BAL, SEX, EDUCATION, MARRIAGE, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6, DEFAULT)

In [45]:
data.describe().select("ID","LIMIT_BAL","SEX","AGE").show(5)

+-----------------+------------------+------------------+-----------------+
|               ID|         LIMIT_BAL|               SEX|              AGE|
+-----------------+------------------+------------------+-----------------+
|            30000|             30000|             30000|            30000|
|          15000.5|167484.32266666667|1.6037333333333332|          35.4855|
|8660.398374208891|129747.66156720246|0.4891291960902602|9.217904068090155|
|                1|             10000|                 1|               21|
|            30000|           1000000|                 2|               79|
+-----------------+------------------+------------------+-----------------+



In [46]:
val renamedData = data.withColumnRenamed("DEFAULT","label");

renamedData = [ID: int, LIMIT_BAL: int ... 23 more fields]


[ID: int, LIMIT_BAL: int ... 23 more fields]

In [47]:
renamedData.select("label").show(5)

+-----+
|label|
+-----+
|    1|
|    1|
|    0|
|    0|
|    0|
+-----+
only showing top 5 rows



In [48]:
val colLabel = "label"
val colNum = renamedData.columns.filter(_ != colLabel)

colLabel = label
colNum = Array(ID, LIMIT_BAL, SEX, EDUCATION, MARRIAGE, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6)


Array(ID, LIMIT_BAL, SEX, EDUCATION, MARRIAGE, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6)

In [49]:
renamedData.select("label").show(5)

+-----+
|label|
+-----+
|    1|
|    1|
|    0|
|    0|
|    0|
+-----+
only showing top 5 rows



In [50]:
for (c <- colNum) {
    println(c)
   val a =renamedData.where(renamedData.col(c).isNull).count()

   println(a) 
}

ID
0
LIMIT_BAL
0
SEX
0
EDUCATION
0
MARRIAGE
0
AGE
0
PAY_0
0
PAY_2
0
PAY_3
0
PAY_4
0
PAY_5
0
PAY_6
0
BILL_AMT1
0
BILL_AMT2
0
BILL_AMT3
0
BILL_AMT4
0
BILL_AMT5
0
BILL_AMT6
0
PAY_AMT1
0
PAY_AMT2
0
PAY_AMT3
0
PAY_AMT4
0
PAY_AMT5
0
PAY_AMT6
0


In [51]:
import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}

val va = new VectorAssembler().setInputCols(colNum).setOutputCol("featuresFlat")
val featuredData = va.transform(renamedData)

val scaler = new StandardScaler().setInputCol("featuresFlat").setOutputCol("scalar")
val scaledData = scaler.fit(featuredData).transform(featuredData)

scaledData.show(5)

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+--------------------+--------------------+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|label|        featuresFlat|              scalar|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+--------------------+--------------------+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0|    1|[1.0,20000.0,2.0,...|[1.15468129385139...|
|  2|   120000|  2|        2|       

va = vecAssembler_2da0039b0c94
featuredData = [ID: int, LIMIT_BAL: int ... 24 more fields]
scaler = stdScal_a56ac9351a15
scaledData = [ID: int, LIMIT_BAL: int ... 25 more fields]


[ID: int, LIMIT_BAL: int ... 25 more fields]

In [52]:
scaledData.columns

Array(ID, LIMIT_BAL, SEX, EDUCATION, MARRIAGE, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6, label, featuresFlat, scalar)

In [53]:
scaledData.select("label").distinct().count()

2

In [54]:
val cols=Array("scalar")
print(cols)

[Ljava.lang.String;@6604b1cf

cols = Array(scalar)


Array(scalar)

In [55]:
// val newData = pipeline.fit(scaledData).transform(scaledData)

val va2 = new VectorAssembler().setInputCols(cols).setOutputCol("features")
val dataset = va2.transform(scaledData).select("features", "label")

va2 = vecAssembler_7d46b23bfa94
dataset = [features: vector, label: int]


[features: vector, label: int]

In [56]:
dataset.select("label").show(5)

+-----+
|label|
+-----+
|    1|
|    1|
|    0|
|    0|
|    0|
+-----+
only showing top 5 rows



In [57]:
val Array(trainSet, testSet) = dataset.randomSplit(Array(0.8, 0.2))

trainSet.show(5)

+--------------------+-----+
|            features|label|
+--------------------+-----+
|(24,[0,1,2,3,4,5,...|    0|
|(24,[0,1,2,3,4,5,...|    0|
|(24,[0,1,2,3,4,5,...|    1|
|(24,[0,1,2,3,4,5,...|    1|
|(24,[0,1,2,3,4,5,...|    1|
+--------------------+-----+
only showing top 5 rows



trainSet = [features: vector, label: int]
testSet = [features: vector, label: int]


[features: vector, label: int]

In [58]:
import org.apache.spark.ml.classification.{BinaryLogisticRegressionSummary, LogisticRegression}
import org.apache.spark.mllib.evaluation.BinaryClassificationMetrics 
import org.apache.spark.ml.evaluation.Evaluator

In [59]:
val lr = new LogisticRegression()
.setFeaturesCol("features").setLabelCol("label")
val lrModel = lr.fit(trainSet)

val trainingSummary = lrModel.summary
print(trainingSummary)

org.apache.spark.ml.classification.BinaryLogisticRegressionTrainingSummaryImpl@5ac37fdd

lr = logreg_89800cc482fc
lrModel = LogisticRegressionModel: uid = logreg_89800cc482fc, numClasses = 2, numFeatures = 24
trainingSummary = org.apache.spark.ml.classification.BinaryLogisticRegressionTrainingSummaryImpl@5ac37fdd


org.apache.spark.ml.classification.BinaryLogisticRegressionTrainingSummaryImpl@5ac37fdd

In [60]:
print(s"The accuracy of the classifier is = ${trainingSummary.accuracy}")

The accuracy of the classifier is = 0.8100913621262459

In [61]:
val predictions = lrModel.transform(testSet)
predictions.show(5)

+--------------------+-----+--------------------+--------------------+----------+
|            features|label|       rawPrediction|         probability|prediction|
+--------------------+-----+--------------------+--------------------+----------+
|(24,[0,1,2,3,4,5,...|    1|[0.76063703527588...|[0.68149202506091...|       0.0|
|(24,[0,1,2,3,4,5,...|    0|[1.26501454083979...|[0.77988812512979...|       0.0|
|(24,[0,1,2,3,4,5,...|    0|[1.07594848893447...|[0.74572650707394...|       0.0|
|(24,[0,1,2,3,4,5,...|    0|[1.05867495812928...|[0.74243724666218...|       0.0|
|(24,[0,1,2,3,4,5,...|    0|[0.96818510686525...|[0.72475760333197...|       0.0|
+--------------------+-----+--------------------+--------------------+----------+
only showing top 5 rows



predictions = [features: vector, label: int ... 3 more fields]


[features: vector, label: int ... 3 more fields]

In [62]:
val lp = predictions.select( "label", "prediction")
val counttotal = predictions.count()
val correct = lp.filter($"label" === $"prediction").count()
val wrong = lp.filter(not($"label" === $"prediction")).count()
val truep = lp.filter($"prediction" === 0.0).filter($"label" === $"prediction").count()
val falseN = lp.filter($"prediction" === 0.0).filter(not($"label" === $"prediction")).count()
val falseP = lp.filter($"prediction" === 1.0).filter(not($"label" === $"prediction")).count()
val ratioWrong=wrong.toDouble/counttotal.toDouble
val ratioCorrect=correct.toDouble/counttotal.toDouble

lp = [label: int, prediction: double]
counttotal = 5920
correct = 4797
wrong = 1123
truep = 4479
falseN = 1013
falseP = 110
ratioWrong = 0.18969594594594594
ratioCorrect = 0.810304054054054


0.810304054054054

In [63]:
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
val evaluator1 = new BinaryClassificationEvaluator().setMetricName("areaUnderROC")
// Evaluates predictions and returns a scalar metric areaUnderROC(larger is better).**
// val evaluator = new BinaryClassificationEvaluator()
val accuracy = evaluator1.evaluate(predictions)
print(accuracy)
// val evaluator = new RegressionEvaluator().setMetricName("areaUnderROC")
// .setPredictionCol("prediction")
// .setLabelCol("label")
// val rmse = evaluator.evaluate(predictions)
// println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

0.7166510122284702

evaluator1 = binEval_5e8481264953
accuracy = 0.7166510122284702


0.7166510122284702

In [64]:
predictions.select("prediction", "label", "features").where("prediction == 0").show(15)

+----------+-----+--------------------+
|prediction|label|            features|
+----------+-----+--------------------+
|       0.0|    1|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    1|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    1|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    1|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
+----------+-----+--------------------+
only showing top 15 rows



In [65]:
val objectiveHistory = trainingSummary.objectiveHistory
objectiveHistory.foreach(loss => println(loss))

0.5272966369818625
0.4836730483045375
0.47913413708286284
0.47830068188874914
0.4739642512745035
0.47019670851094714
0.465899269083548
0.46560632204846114
0.46488666912752896
0.4647789000291914
0.4646365390655921
0.46450610082117777
0.4643777851040914
0.4643220713897213
0.46429359956426536
0.46427850151385563
0.4642481615564603
0.4642255750686248
0.46419212638715707
0.46415595829210426
0.4640873000363526
0.46405523492633505
0.46401047176093213
0.4639910215941
0.46397176090796344
0.4639687994148078
0.4639561119834347
0.46394812348773506
0.4639413277601859
0.4639373931472524
0.4639319873034502
0.4639256775736155
0.4639216214895904
0.4639188201544619
0.463915801762874
0.46391001624256706
0.4639094237630225
0.46390708496060157
0.4639056983116319
0.4639035788848467
0.46390196766992714
0.4639014403003657
0.4639006537611187
0.4639002882383104
0.4639001769540668
0.46390015173910526
0.46390002830437355
0.46390001534417863
0.46389997656619086
0.4638999262553854
0.4638998902090737
0.4638998327208

objectiveHistory = Array(0.5272966369818625, 0.4836730483045375, 0.47913413708286284, 0.47830068188874914, 0.4739642512745035, 0.47019670851094714, 0.465899269083548, 0.46560632204846114, 0.46488666912752896, 0.4647789000291914, 0.4646365390655921, 0.46450610082117777, 0.4643777851040914, 0.4643220713897213, 0.46429359956426536, 0.46427850151385563, 0.4642481615564603, 0.4642255750686248, 0.46419212638715707, 0.46415595829210426, 0.4640873000363526, 0.46405523492633505, 0.46401047176093213, 0.4639910215941, 0.46397176090796344, 0.4639687994148078, 0.4639561119834347, 0.46394812348773506, 0.4639413277601859, 0.4639373931472524, 0.4639319873034502, 0.4639256775736155, 0.4639216214895904, 0.4639188201544619, 0.463915801762874, 0.46391001624256706, 0.4639094237630225, 0.46390...


Array(0.5272966369818625, 0.4836730483045375, 0.47913413708286284, 0.47830068188874914, 0.4739642512745035, 0.47019670851094714, 0.465899269083548, 0.46560632204846114, 0.46488666912752896, 0.4647789000291914, 0.4646365390655921, 0.46450610082117777, 0.4643777851040914, 0.4643220713897213, 0.46429359956426536, 0.46427850151385563, 0.4642481615564603, 0.4642255750686248, 0.46419212638715707, 0.46415595829210426, 0.4640873000363526, 0.46405523492633505, 0.46401047176093213, 0.4639910215941, 0.46397176090796344, 0.4639687994148078, 0.4639561119834347, 0.46394812348773506, 0.4639413277601859, 0.4639373931472524, 0.4639319873034502, 0.4639256775736155, 0.4639216214895904, 0.4639188201544619, 0.463915801762874, 0.46391001624256706, 0.4639094237630225, 0.46390...

In [66]:
// Obtain the metrics useful to judge performance on test data.
// We cast the summary to a BinaryLogisticRegressionSummary since the problem is a
// binary classification problem.
val binarySummary = trainingSummary.asInstanceOf[BinaryLogisticRegressionSummary]

// Obtain the receiver-operating characteristic as a dataframe and areaUnderROC.
val roc = binarySummary.roc
roc.show()
println(binarySummary.areaUnderROC)

// Set the model threshold to maximize F-Measure
val fMeasure = binarySummary.fMeasureByThreshold
val maxFMeasure = fMeasure.select(max("F-Measure")).head().getDouble(0)
val bestThreshold = fMeasure.where($"F-Measure" === maxFMeasure)
  .select("threshold").head().getDouble(0)
lrModel.setThreshold(bestThreshold)

+--------------------+-------------------+
|                 FPR|                TPR|
+--------------------+-------------------+
|                 0.0|                0.0|
|0.003462050599201...|0.03298774740810556|
|0.007456724367509...| 0.0640904806786051|
| 0.01081225033288948|0.09745523091423186|
|0.014007989347536618|0.13138548539114042|
|0.017949400798934752|0.16267672007540057|
|0.021784287616511317| 0.1943449575871819|
|0.025459387483355526|0.22657869934024505|
| 0.03062583222370173| 0.2535344015080113|
|0.035525965379494005| 0.2814326107445806|
| 0.04047936085219707| 0.3091423185673893|
| 0.04585885486018642| 0.3353440150801131|
| 0.05171770972037284| 0.3598491988689915|
| 0.05821571238348868| 0.3820923656927427|
| 0.06545938748335553|0.40169651272384543|
| 0.07344873501997337| 0.4186616399622997|
| 0.08154460719041279| 0.4352497643732328|
| 0.09049267643142477| 0.4488218661639962|
| 0.09880159786950732| 0.4646559849198869|
| 0.10737683089214381|0.47954759660697455|
+----------

binarySummary = org.apache.spark.ml.classification.BinaryLogisticRegressionTrainingSummaryImpl@5ac37fdd
roc = [FPR: double, TPR: double]
fMeasure = [threshold: double, F-Measure: double]
maxFMeasure = 0.5187046882551958
bestThreshold = 0.27584584603559487


LogisticRegressionModel: uid = logreg_89800cc482fc, numClasses = 2, numFeatures = 24

# Decision Tree

In [67]:
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
import org.apache.spark.ml.classification.DecisionTreeClassifier

val dt = new DecisionTreeClassifier().setLabelCol("label").setFeaturesCol("features")

// train the model
val dtModel = dt.fit(trainSet)

// make predictions on the test data
val predictions = dtModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
// val evaluator = new RegressionEvaluator().setMetricName("rmse")
// .setPredictionCol("prediction")
// .setLabelCol("label")
// val rmse = evaluator.evaluate(predictions)
// println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+----------+-----+--------------------+
|prediction|label|            features|
+----------+-----+--------------------+
|       0.0|    1|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
+----------+-----+--------------------+
only showing top 5 rows



dt = dtc_39a396c02f9a
dtModel = DecisionTreeClassificationModel (uid=dtc_39a396c02f9a) of depth 5 with 41 nodes
predictions = [features: vector, label: int ... 3 more fields]


[features: vector, label: int ... 3 more fields]

In [68]:
val evaluator = new BinaryClassificationEvaluator()
val accuracy = evaluator.evaluate(predictions)

print(accuracy)

0.3211521229923122

evaluator = binEval_97de3e5c7c09
accuracy = 0.3211521229923122


0.3211521229923122

In [69]:
val lp = predictions.select( "label", "prediction")
val counttotal = predictions.count()
val correct = lp.filter($"label" === $"prediction").count()
val wrong = lp.filter(not($"label" === $"prediction")).count()
val truep = lp.filter($"prediction" === 0.0).filter($"label" === $"prediction").count()
val falseN = lp.filter($"prediction" === 0.0).filter(not($"label" === $"prediction")).count()
val falseP = lp.filter($"prediction" === 1.0).filter(not($"label" === $"prediction")).count()
val ratioWrong=wrong.toDouble/counttotal.toDouble
val ratioCorrect=correct.toDouble/counttotal.toDouble

lp = [label: int, prediction: double]
counttotal = 5920
correct = 4840
wrong = 1080
truep = 4382
falseN = 873
falseP = 207
ratioWrong = 0.18243243243243243
ratioCorrect = 0.8175675675675675


0.8175675675675675

# Random Forest

In [70]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.classification.RandomForestClassifier

val rf = new RandomForestClassifier().setLabelCol("label").setFeaturesCol("features").setNumTrees(10)

// train the model
val rfModel = rf.fit(trainSet)

//make predictions on the test data
val predictions = rfModel.transform(testSet)
predictions.select("prediction", "label", "features").show(15)

// select (prediction, true label) and compute test error
// val evaluator = new RegressionEvaluator().setMetricName("rmse")
// .setPredictionCol("prediction")
// .setLabelCol("label")
// val rmse = evaluator.evaluate(predictions)
// println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+----------+-----+--------------------+
|prediction|label|            features|
+----------+-----+--------------------+
|       0.0|    1|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    1|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    1|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    1|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
|       0.0|    0|(24,[0,1,2,3,4,5,...|
+----------+-----+--------------------+
only showing top 15 rows



rf = rfc_7d674a38cd22
rfModel = RandomForestClassificationModel (uid=rfc_7d674a38cd22) with 10 trees
predictions = [features: vector, label: int ... 3 more fields]


[features: vector, label: int ... 3 more fields]

In [71]:
predictions.select("prediction", "label", "features").where("prediction == 1").show(15)

+----------+-----+--------------------+
|prediction|label|            features|
+----------+-----+--------------------+
|       1.0|    0|(24,[0,1,2,3,4,5,...|
|       1.0|    1|(24,[0,1,2,3,4,5,...|
|       1.0|    0|(24,[0,1,2,3,4,5,...|
|       1.0|    1|(24,[0,1,2,3,4,5,...|
|       1.0|    1|(24,[0,1,2,3,4,5,...|
|       1.0|    1|(24,[0,1,2,3,4,5,...|
|       1.0|    1|(24,[0,1,2,3,4,5,...|
|       1.0|    1|(24,[0,1,2,3,4,5,...|
|       1.0|    1|(24,[0,1,2,3,4,5,...|
|       1.0|    0|(24,[0,1,2,3,4,5,...|
|       1.0|    0|(24,[0,1,2,3,4,5,...|
|       1.0|    1|(24,[0,1,2,3,4,5,...|
|       1.0|    1|[0.00265576697585...|
|       1.0|    1|[0.00727449215126...|
|       1.0|    0|[0.00808276905695...|
+----------+-----+--------------------+
only showing top 15 rows



In [72]:
val evaluator = new BinaryClassificationEvaluator()
val accuracy = evaluator.evaluate(predictions)

print(accuracy)

evaluator.getMetricName()

Name: Compile Error
Message: <console>:140: error: not enough arguments for method apply: (index: Int)Char in class StringOps.
Unspecified value parameter index.
       evaluator.getMetricName()
                              ^

StackTrace: 

In [73]:
val lp = predictions.select( "label", "prediction")
val counttotal = predictions.count()
val correct = lp.filter($"label" === $"prediction").count()
val wrong = lp.filter(not($"label" === $"prediction")).count()
val truep = lp.filter($"prediction" === 0.0).filter($"label" === $"prediction").count()
val falseN = lp.filter($"prediction" === 0.0).filter(not($"label" === $"prediction")).count()
val falseP = lp.filter($"prediction" === 1.0).filter(not($"label" === $"prediction")).count()
val ratioWrong=wrong.toDouble/counttotal.toDouble
val ratioCorrect=correct.toDouble/counttotal.toDouble

lp = [label: int, prediction: double]
counttotal = 5920
correct = 4828
wrong = 1092
truep = 4393
falseN = 896
falseP = 196
ratioWrong = 0.18445945945945946
ratioCorrect = 0.8155405405405406


0.8155405405405406